### PR TITLE
fix user enumeration with logging

### PR DESCRIPTION
**Pull request by peterprochaska**
## Description
Fixed user enumeration

## Related Issue
https://github.com/owncloud/core/issues/23595

## Motivation and Context
instead of exceptions, we are logging

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: core/Controller/LostController.php [hunk 1]
```diff
@@ -236,42 +242,41 @@ public function setPassword($token, $userId, $password, $proceed) {
...
-		$this->config->setUserValue($user, 'owncloud', 'lostpassword', $this->timeFactory->getTime() .':'. $token);
+			if (!empty($email)) {
+				$token = $this->secureRandom->generate(21,
+					ISecureRandom::CHAR_DIGITS .
+					ISecureRandom::CHAR_LOWER .
+					ISecureRandom::CHAR_UPPER);
+				$this->config->setUserValue($user, 'owncloud', 'lostpassword', $this->timeFactory->getTime() . ':' . $token);
 
-		$link = $this->urlGenerator->linkToRouteAbsolute('core.lost.resetform', ['userId' => $user, 'token' => $token]);
+				$link = $this->urlGenerator->linkToRouteAbsolute('core.lost.resetform', ['userId' => $user, 'token' => $token]);
```
Comment (PVince81) [inline core/Controller/LostController.php]: weird indent, please use tabs

File: core/Controller/LostController.php [hunk 2]
```diff
@@ -236,42 +242,41 @@ public function setPassword($token, $userId, $password, $proceed) {
...
+					$message->setPlainBody($msg);
+					$message->setFrom([$this->from => $this->defaults->getName()]);
+					$this->mailer->send($message);
+				} catch (\Exception $e) {
+					throw new \Exception($this->l10n->t(
+						'Couldn\'t send reset email. Please contact your administrator.'
+					));
+				}
+			} else {
+				$this->logger->error('Could not send reset email because there is no email address for this username.', ['app' => 'core']);
```
Comment (PVince81) [inline core/Controller/LostController.php]: please add the user in the log message, could be useful for fail2ban or so

File: core/Controller/LostController.php [hunk 3]
```diff
@@ -236,42 +242,41 @@ public function setPassword($token, $userId, $password, $proceed) {
...
+			}
+		} else {
+			$this->logger->error('Couldn\'t send reset email because User does not exist.', ['app' => 'core']);
```
Comment (PVince81) [inline core/Controller/LostController.php]: same here, add the user name in the message

File: core/Controller/LostController.php [hunk 4]
```diff
@@ -236,42 +242,41 @@ public function setPassword($token, $userId, $password, $proceed) {
...
-		}
+			$userObject = $this->userManager->get($user);
+			$email = $userObject->getEMailAddress();
 
-		$token = $this->secureRandom->generate(21,
-			ISecureRandom::CHAR_DIGITS.
-			ISecureRandom::CHAR_LOWER.
-			ISecureRandom::CHAR_UPPER);
-		$this->config->setUserValue($user, 'owncloud', 'lostpassword', $this->timeFactory->getTime() .':'. $token);
+			if (!empty($email)) {
```
Comment (PVince81) [inline core/Controller/LostController.php]: Note that in PHP `empty("0")` is true. Usually that won't happen with email addresses.

You might want to change it but I'm fine if you don't as it's part of the old code.

**Review by PVince81 — COMMENTED** [review]
Code changes look good, I do miss the user name in the log.

**Comment by PVince81** [comment]
@Peter-Prochaska please also adjust the related unit tests, they are failing.

File: core/Controller/LostController.php (hunk 4 above)
Comment (jvillafanez) [inline core/Controller/LostController.php]: The `getEMailAddress()` function should have documented what is the expected result if there is no email set (maybe null?). I'd make the condition to match that expectation.

File: core/Controller/LostController.php [hunk 5]
```diff
@@ -236,42 +242,41 @@ public function setPassword($token, $userId, $password, $proceed) {
...
 
-		$link = $this->urlGenerator->linkToRouteAbsolute('core.lost.resetform', ['userId' => $user, 'token' => $token]);
+				$link = $this->urlGenerator->linkToRouteAbsolute('core.lost.resetform', ['userId' => $user, 'token' => $token]);
 
-		$tmpl = new \OC_Template('core', 'lostpassword/email');
-		$tmpl->assign('link', $link);
-		$msg = $tmpl->fetchPage();
+				$tmpl = new \OC_Template('core', 'lostpassword/email');
+				$tmpl->assign('link', $link);
+				$msg = $tmpl->fetchPage();
```
Comment (PVince81) [inline core/Controller/LostController.php]: please use tabs, not spaces

**Comment by PVince81** [comment]
@Peter-Prochaska thanks for fixing the unit tests. Can you also address the review comments ?

**Comment by PVince81** [comment]
Retested, works :+1:

Let's merge after Jenkins passes then move on to the backports.

**Comment by PVince81** [comment]
Travis is unlikely to fail, but due to tech issues on their side it might take a while to catch up.

Since Jenkins passed it is very likely that it will all be green.

@Peter-Prochaska please prepare the backport PRs to all relevant branches.

**Comment by PVince81** [comment]
Merged.

@Peter-Prochaska please backport

**Comment by PVince81** [comment]
stable9.1: https://github.com/owncloud/core/pull/27021
stable9: https://github.com/owncloud/core/pull/27025
stable8.2: https://github.com/owncloud/core/pull/27029
stable8.1: https://github.com/owncloud/core/pull/27031

**Pull request merged**